### ISSUE TITLE
ArrayDeclaration usefulness nuked by excludes

### ISSUE THREAD

**Issue by Majkl578** [squizlabs/PHP_CodeSniffer#1991]
Consider the following ruleset:
```xml
<?xml version="1.0"?>
<ruleset name="bug">
    <rule ref="Squiz.Arrays.ArrayDeclaration">
        <exclude name="Squiz.Arrays.ArrayDeclaration.SingleLineNotAllowed"/>
        <exclude name="Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned"/>
        <exclude name="Squiz.Arrays.ArrayDeclaration.KeyNotAligned"/>
    </rule>

    <file>test.php</file>
</ruleset>
```
The purpose here is to have `=>` aligned, but not the keys or closing brace.

So the following code is valid:
```php
<?php

$a = [1 => 1, 2 => 2];
$b = [
    'zero' => 0,
    'one'  => 1,
];
```

Now let's break the alignment for `=>`:
```php
<?php

$a = [1 => 1, 2 => 2];
$b = [
    'zero' => 0,
    'one' => 1,
];
```

Now run phpcs: voila, no error!
```
$ vendor/bin/phpcs test.php -sp
. 1 / 1 (100%)


Time: 21ms; Memory: 4Mb
```

Wait, what? The alignment is obviously broken, right? But not reported.
Here's the gotcha: if I accommodate the needs of `Squiz.Arrays.ArrayDeclaration.KeyNotAligned`, the error suddenly appears!
So:
```php
<?php

$a = [1 => 1, 2 => 2];
$b = [
      'zero' => 0,
      'one' => 1,
];
```
Correctly reports:
```
$ vendor/bin/phpcs test.php -sp
E 1 / 1 (100%)



FILE: /tmp/broken-phpcs/test.php
-------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-------------------------------------------------------------------------------------------------------------------------------------------------
 6 | ERROR | [x] Array double arrow not aligned correctly; expected 2 space(s) but found 1 (Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned)
-------------------------------------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-------------------------------------------------------------------------------------------------------------------------------------------------

Time: 27ms; Memory: 4Mb
```

Now let's see why this happens. The source is checking the errors one by one in a loop. When an error is detected, the iteration is aborted and immediately skips to the next index. Even if the error is being ignored!
https://github.com/squizlabs/PHP_CodeSniffer/blob/f58ea162aac228d7f336d25216ca1613030da4a5/src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php#L665-L866

So using `continue` here without checking whether the error is actually enabled is to blame here.

_(Feel free to retitle to something more explicit.)_

**Comment by jrfnl** [comment]
Duplicate of #582 ?

**Comment by gsherwood** [comment]
Yes, I know this sniff is pretty useless at being used to cherry pick individual rules. It simple contains too many, and became too complex.

The solution I see (as outlined in #582) is to break this up into multiple sniffs so people can pick just the checks they want.

This started with the new `Generic.Arrays.ArrayIndent` sniff and the creation of `AbstractArraySniff` class so that array-based sniffs can be more easily created.

Follow along on that issues if you want to track progress. I've been busy with releases lately, so I haven't had time to move any more of the rules out into the Generic standard, but the "arrow alignment" and "comma after last" checks are likely next on the list as I imagine they are the most common.

**Comment by gsherwood** [comment]
Closing so discussion can continue on #582
